File: MiniDB_Projects/Team_DBAlchemist/src/catalog/catalog.py
"""
Catalog: stores table schemas (column names, types, primary key).
Persisted as JSON. Loaded on startup.

Column types supported: INT, TEXT, FLOAT
"""
import json
import os


SUPPORTED_TYPES = {'INT', 'TEXT', 'FLOAT'}


class Column:
    def __init__(self, name: str, col_type: str, primary_key: bool = False):
        self.name = name
        self.col_type = col_type.upper()
        self.primary_key = primary_key

    def to_dict(self) -> dict:
        return {'name': self.name, 'type': self.col_type, 'primary_key': self.primary_key}

    @staticmethod
    def from_dict(d: dict) -> 'Column':
        return Column(d['name'], d['type'], d.get('primary_key', False))

    def cast(self, value):
        """Cast a string value to this column's Python type."""
        if value is None:
            return None
        if self.col_type == 'INT':
            return int(value)
        if self.col_type == 'FLOAT':
            return float(value)
        return str(value)


class TableSchema:
    def __init__(self, name: str, columns: list[Column]):
        self.name = name
        self.columns = columns
        self._col_map = {c.name: c for c in columns}

    def get_column(self, name: str) -> Column | None:
        return self._col_map.get(name)

    def primary_key_col(self) -> Column | None:
        for c in self.columns:
            if c.primary_key:
                return c
        return None

    def col_names(self) -> list[str]:
        return [c.name for c in self.columns]

    def to_dict(self) -> dict:
        return {'name': self.name, 'columns': [c.to_dict() for c in self.columns]}

    @staticmethod
    def from_dict(d: dict) -> 'TableSchema':
        cols = [Column.from_dict(c) for c in d['columns']]
        return TableSchema(d['name'], cols)

    def cast_row(self, row: dict) -> dict:
        """Cast all values in a row dict to correct Python types."""
        return {k: self._col_map[k].cast(v) if k in self._col_map else v
                for k, v in row.items()}


class Catalog:
    def __init__(self, path: str):
        self.path = path
        self._tables: dict[str, TableSchema] = {}
        self._load()

    def create_table(self, name: str, columns: list[Column]):
        if name in self._tables:
            raise ValueError(f"Table '{name}' already exists")
        schema = TableSchema(name, columns)
        self._tables[name] = schema
        self._save()
        return schema

    def get_table(self, name: str) -> TableSchema | None:
        return self._tables.get(name)

    def drop_table(self, name: str):
        if name not in self._tables:
            raise ValueError(f"Table '{name}' does not exist")
        del self._tables[name]
        self._save()

    def list_tables(self) -> list[str]:
        return list(self._tables.keys())

    def _save(self):
        data = {name: schema.to_dict() for name, schema in self._tables.items()}
        with open(self.path, 'w') as f:
            json.dump(data, f, indent=2)

    def _load(self):
        if not os.path.exists(self.path):
            return
        with open(self.path, 'r') as f:
            data = json.load(f)
        self._tables = {name: TableSchema.from_dict(d) for name, d in data.items()}

File: MiniDB_Projects/Team_XO/internal/catalog/catalog.go
// Package catalog holds table metadata (schemas) and owns the encoding of rows
// into the opaque byte tuples that the storage engine persists. Centralising the
// encoding here means the storage layer never needs to understand column types,
// and every other layer agrees on a single on-disk record format.
package catalog

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
	"os"
	"sync"

	"minidb/internal/types"
)

// Column is a named, typed attribute of a table.
type Column struct {
	Name string           `json:"name"`
	Type types.ColumnType `json:"type"`
}

// Table describes one relation: its columns and which column is the primary
// key. The primary key column is automatically indexed by the engine.
type Table struct {
	Name      string   `json:"name"`
	Columns   []Column `json:"columns"`
	PKColumn  int      `json:"pk_column"`  // index into Columns; -1 if none
	HasSecond bool     `json:"has_second"` // whether a secondary index exists
	SecondCol int      `json:"second_col"` // column index for the secondary index
}

// ColumnIndex returns the position of the named column, or -1 if absent.
func (t *Table) ColumnIndex(name string) int {
	for i, c := range t.Columns {
		if c.Name == name {
			return i
		}
	}
	return -1
}

// EncodeRow serialises a row into the heap-file tuple format. The layout is a
// sequence of per-column fields:
//
//	INT  : 1 null byte + 8 bytes little-endian
//	BOOL : 1 null byte + 1 byte
//	TEXT : 1 null byte + 4 byte length + raw UTF-8 bytes
//
// A leading null byte per column keeps the format self-describing for NULLs
// without a separate null bitmap, which is plenty for MiniDB's scale.
func (t *Table) EncodeRow(row types.Row) []byte {
	var buf []byte
	for i, col := range t.Columns {
		v := row[i]
		if v.Null {
			buf = append(buf, 1)
		} else {
			buf = append(buf, 0)
		}
		switch col.Type {
		case types.TypeInt:
			var b [8]byte
			binary.LittleEndian.PutUint64(b[:], uint64(v.Int))
			buf = append(buf, b[:]...)
		case types.TypeBool:
			if v.Bool {
				buf = append(buf, 1)
			} else {
				buf = append(buf, 0)
			}
		case types.TypeText:
			var b [4]byte
			binary.LittleEndian.PutUint32(b[:], uint32(len(v.Str)))
			buf = append(buf, b[:]...)
			buf = append(buf, []byte(v.Str)...)
		}
	}
	return buf
}

// DecodeRow reverses EncodeRow using the table schema to drive parsing.
func (t *Table) DecodeRow(b []byte) types.Row {
	row := make(types.Row, len(t.Columns))
	pos := 0
	for i, col := range t.Columns {
		isNull := b[pos] == 1
		pos++
		switch col.Type {
		case types.TypeInt:
			v := int64(binary.LittleEndian.Uint64(b[pos:]))
			pos += 8
			row[i] = types.Value{Type: types.TypeInt, Int: v, Null: isNull}
		case types.TypeBool:
			v := b[pos] == 1
			pos++
			row[i] = types.Value{Type: types.TypeBool, Bool: v, Null: isNull}
		case types.TypeText:
			n := int(binary.LittleEndian.Uint32(b[pos:]))
			pos += 4
			s := string(b[pos : pos+n])
			pos += n
			row[i] = types.Value{Type: types.TypeText, Str: s, Null: isNull}
		}
	}
	return row
}

// Catalog is the in-memory registry of all tables, with simple JSON persistence
// so schemas survive restarts and are available to crash recovery.
type Catalog struct {
	mu     sync.RWMutex
	path   string
	Tables map[string]*Table `json:"tables"`
}

// NewCatalog loads an existing catalog from path, or starts an empty one if the
// file does not exist yet.
func NewCatalog(path string) (*Catalog, error) {
	c := &Catalog{path: path, Tables: make(map[string]*Table)}
	data, err := os.ReadFile(path)
	if os.IsNotExist(err) {
		return c, nil
	}
	if err != nil {
		return nil, err
	}
	if err := json.Unmarshal(data, &struct {
		Tables *map[string]*Table `json:"tables"`
	}{Tables: &c.Tables}); err != nil {
		return nil, fmt.Errorf("catalog: parse %q: %w", path, err)
	}
	return c, nil
}

// CreateTable registers a new table and persists the catalog.
func (c *Catalog) CreateTable(t *Table) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	if _, ok := c.Tables[t.Name]; ok {
		return fmt.Errorf("catalog: table %q already exists", t.Name)
	}
	c.Tables[t.Name] = t
	return c.saveLocked()
}

// Get returns the table by name.
func (c *Catalog) Get(name string) (*Table, bool) {
	c.mu.RLock()
	defer c.mu.RUnlock()
	t, ok := c.Tables[name]
	return t, ok
}

// List returns all table names (unordered).
func (c *Catalog) List() []string {
	c.mu.RLock()
	defer c.mu.RUnlock()
	out := make([]string, 0, len(c.Tables))
	for name := range c.Tables {
		out = append(out, name)
	}
	return out
}

// Save persists the catalog after an in-place mutation of a table (such as
// adding a secondary index).
func (c *Catalog) Save() error {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.saveLocked()
}

func (c *Catalog) saveLocked() error {
	if c.path == "" {
		return nil
	}
	data, err := json.MarshalIndent(struct {
		Tables map[string]*Table `json:"tables"`
	}{Tables: c.Tables}, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(c.path, data, 0o644)
}
